From fa6f101731c260856b01cb7aeb978333a89d43d4 Mon Sep 17 00:00:00 2001
From: Nidhi Jain <quic_nidhjain@quicinc.com>
Date: Wed, 28 Aug 2024 15:24:53 +0530
Subject: [PATCH] cfg80211: Add 4.9GHz support in freq_reg_info

The AP is not able to come up on 4.9GHz channels 4942, 4945, 4947,
4982, 4985 and 4987 MHz.

Since 20 MHz is the minimum bandwidth configured in freq_reg_info,
cfg80211_does_bw_fit_range fails to return true for the above frequencies.

To fix this issue, in cfg80211_does_bw_fit_range, if center_freq is
4.9GHz, then do not modify the start_freq and end_freq

Change-Id: I78060da39d982cc435d06ce1d49e67d02dab3e2b
Signed-off-by: Nidhi Jain <quic_nidhjain@quicinc.com>
---
 net/wireless/util.c | 15 +++++++++++++--
 1 file changed, 13 insertions(+), 2 deletions(-)

diff --git a/net/wireless/util.c b/net/wireless/util.c
index ecb42589a400..2ad12a71c0ad 100644
--- a/net/wireless/util.c
+++ b/net/wireless/util.c
@@ -2435,8 +2435,19 @@ bool cfg80211_does_bw_fit_range(const struct ieee80211_freq_range *freq_range,
 {
 	u32 start_freq_khz, end_freq_khz;
 
-	start_freq_khz = center_freq_khz - (bw_khz / 2);
-	end_freq_khz = center_freq_khz + (bw_khz / 2);
+	/* As 4.9GHz supports 5Mhz and 10 MHz center frequencies,
+	 * the offset calculation using the bw_khz may not work.
+	 * Therefore, apply center_freq_khz to start_freq_khz and
+	 * end_freq_khz directly for bw check.
+	 */
+	if (center_freq_khz >= MHZ_TO_KHZ(4940) &&
+	    center_freq_khz <= MHZ_TO_KHZ(5090)) {
+		start_freq_khz = center_freq_khz;
+		end_freq_khz = center_freq_khz;
+	} else {
+		start_freq_khz = center_freq_khz - (bw_khz / 2);
+		end_freq_khz = center_freq_khz + (bw_khz / 2);
+	}
 
 	if (start_freq_khz >= freq_range->start_freq_khz &&
 	    end_freq_khz <= freq_range->end_freq_khz)
-- 
2.34.1

